src/faidx/mod.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2019 Manuel Landesfeind, Evotec International GmbH
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//!
+//! Module for working with faidx-indexed FASTA files.
+//!
+
+use std::ffi;
+use std::path::Path;
+use url::Url;
+
+use crate::htslib;
+
+pub mod errors;
+pub use errors::{Error, Result};
+
+fn path_as_bytes<'a, P: 'a + AsRef<Path>>(path: P, must_exist: bool) -> Result<Vec<u8>> {
+    if path.as_ref().exists() || !must_exist {
+        Ok(path
+            .as_ref()
+            .to_str()
+            .ok_or(Error::NonUnicodePath)?
+            .as_bytes()
+            .to_owned())
+    } else {
+        Err(Error::FileNotFound {
+            path: path.as_ref().to_owned(),
+        })
+    }
+}
+/// A Fasta reader.
+#[derive(Debug)]
+pub struct Reader {
+    inner: *mut htslib::faidx_t,
+}
+
+impl Reader {
+    /// Create a new Reader from a path.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path to open.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
+        Self::new(&path_as_bytes(path, true)?)
+    }
+
+    /// Create a new Reader from an URL.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - the url to open
+    pub fn from_url(url: &Url) -> Result<Self, Error> {
+        Self::new(url.as_str().as_bytes())
+    }
+
+    /// Internal function to create a Reader from some sort of path (could be file path but also URL).
+    /// The path or URL will be handled by the c-implementation transparently.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path or URL to open
+    fn new(path: &[u8]) -> Result<Self, Error> {
+        let cpath = ffi::CString::new(path).unwrap();
+        let inner = unsafe { htslib::fai_load(cpath.as_ptr()) };
+        Ok(Self { inner })
+    }
+
+    /// Fetches the sequence and returns it.
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - the name of the template sequence (e.g., "chr1")
+    /// * `begin` - the offset within the template sequence (starting with 0)
+    /// * `end` - the end position to return (if smaller than `begin`, the behavior is undefined).
+    pub fn fetch_seq<N: AsRef<str>>(&self, name: N, begin: usize, end: usize) -> String {
```

Comment:
I think the bounds should be u64. usize can be 32 bit on 32 bit machines, which might be not enough to index into a large fasta file.

Reply:
Is there a particular reason to return a unicode string? Why not just bytes?

Reply:
Also, I think it might make sense to let people provide a buffer, instead of returning a newly allocated string every time.

Reply:
Hi Johannes, thanks a lot for the suggestions!

> I think the bounds should be u64. usize can be 32 bit on 32 bit machines, which might be not enough to index into a large fasta file.

Valid point. The internal htslib type is c_int (=i32). Either we use that directly or switch to `faidx_fetch_seq64()` using hts_pos_t (=i64). I would suggest the latter but its your call :-)

> Is there a particular reason to return a unicode string? Why not just bytes?

Because I like Strings over raw byte arrays (because I am a simple mind and did not need optimize my code for speed so far). But clearly can change that. I will implement this along with a pre-allocated buffer. Maybe I that as separate functions (i.e., syntactic sugar). Just give me a couple days, because I have a busy week.

Reply:
Hi @johanneskoester, 1770a09e437558a96cce4bbcc2eab2c011d8b96e updates the branch to use i64 type as parameter. This is consistent with the htslib C-API that uses [hts_pos_t](https://github.com/samtools/htslib/blob/835e133b941a175246b18a1a3261f7d91611a1dc/faidx.c#L695) which is [defined as i64](https://github.com/samtools/htslib/blob/34c15969fa49d78b07c649374fb008fb3ed1e98a/htslib/hts.h#L629). I agree that `u64` would be nicer but we then would need to check overflows and either return a `Result` or `panic` - both options I do not like.

I changed the return value to byte array. Providing a buffer is not useful because the C-API creates a new byte array anyways and [converting CStr to bytes should be computationally cheap](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_bytes).

Reply:
Why is `Result` not acceptable as a return type?

Reply:
`Result` is a perfectly acceptable return type in general but in this case the single thing that can fail is the cast. A developer using this API needs take care either before (i.e., doing the cast by him/herself) or after (i.e., matching the Result) the call. Therefore, I consider returning a `Result` here as a kind of overhead.

I also think that @lh3 (and/or htslib-team) had a reason to choose `i64` as type for positions. Imho the `rust-htslib` wrappers should resemble the `C-htslib` API closely. Even if I don't know what `C-htslib` will do if a negative value is supplied, one day a developer might want to use the `rust-htslib` in that way.

Reply:
See https://github.com/samtools/hts-specs/pull/460 for a good discussion on signedness (for .BAI format in this case but the rationale should apply to the rest of positional-encoding fields).

TL;DR: positions should be positive on the current spec, if they are not, it's probably a bug in a spec that hasn't seen proper scrutiny in ~10 years ;)

Reply:
Hi @brainstorm, 
I agree that making the API ergonomic for Rust programmers is a good goal. Also, I was not aware of the discussion in hts-specs and naively assumed that i64 was used for a specific reason. I now switched to `usize` type and added an Error type for in case conversion fails. Not sure though if the description gives enough detail:
```
#[derive(Error, Debug, PartialEq)]
pub enum Error {
   ...
    #[error("The given position is to large to be converted to i64")]
    PositionTooLarge,
}